core/drivers/imx_ele.c

```diff
@@ -0,0 +1,576 @@
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * Open a session with EdgeLock Enclave. It return a session handler.
+ *
+ * @session_handle EdgeLock Enclave session handler
+ */
+static TEE_Result imx_ele_session_open(uint32_t *session_handle)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct open_session_cmd {
+		uint8_t mu_id;
+		uint8_t interrupt_num;
+		uint8_t tz;
+		uint8_t did;
+		uint8_t priority;
+		uint8_t op_mode;
+		uint16_t reserved;
+	} __packed cmd = {
+		.mu_id = ELE_MU_ID,
+		.interrupt_num = ELE_MU_IRQ,
+		.tz = 0,
+		.did = ELE_MU_DID,
+		.priority = 0,
+		.op_mode = 0,
+		.reserved = 0,
+	};
+	struct open_session_rsp {
+		uint32_t rsp_code;
+		uint32_t session_handle;
+	} rsp = { };
+	struct imx_mu_msg msg = {
+		.header.version = ELE_VERSION_HSM,
+		.header.size = SIZE_MSG_32(cmd),
+		.header.tag = ELE_REQUEST_TAG,
+		.header.command = ELE_CMD_SESSION_OPEN,
+	};
+
+	assert(session_handle);
+
+	memcpy(msg.data.u8, &cmd, sizeof(cmd));
+
+	res = imx_ele_call(&msg);
+	if (res)
+		return res;
+
+	memcpy(&rsp, msg.data.u8, sizeof(rsp));
+
+	if (session_handle)
+		*session_handle = rsp.session_handle;
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * Close a session with EdgeLock Enclave.
+ *
+ * @session_handle EdgeLock Enclave session handler
+ */
+static TEE_Result imx_ele_session_close(uint32_t session_handle)
+{
+	struct close_session_cmd {
+		uint32_t session_handle;
+	} cmd = {
+		.session_handle = session_handle,
+	};
+	struct imx_mu_msg msg = {
+		.header.version = ELE_VERSION_HSM,
+		.header.size = SIZE_MSG_32(cmd),
+		.header.tag = ELE_REQUEST_TAG,
+		.header.command = ELE_CMD_SESSION_CLOSE,
+	};
+
+	memcpy(msg.data.u8, &cmd, sizeof(cmd));
+
+	return imx_ele_call(&msg);
+}
+
+/*
+ * Get the current state of the ELE TRNG
+ */
+static TEE_Result imx_ele_rng_get_trng_state(void)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct rng_get_trng_state_msg_rsp {
+		uint32_t rsp_code;
+		uint8_t trng_state;
+		uint8_t csal_state;
+	} __packed rsp = { };
+	struct imx_mu_msg msg = {
+		.header.version = ELE_VERSION_BASELINE,
+		.header.size = 1,
+		.header.tag = ELE_REQUEST_TAG,
+		.header.command = ELE_CMD_TRNG_STATE,
+	};
+
+	res = imx_ele_call(&msg);
+	if (res)
+		return res;
+
+	memcpy(&rsp, msg.data.u8, sizeof(rsp));
+
+	if (rsp.trng_state != IMX_ELE_TRNG_STATUS_READY)
+		return TEE_ERROR_BUSY;
+	else
+		return TEE_SUCCESS;
+}
+
+/*
+ * Get random data from the EdgeLock Enclave
+ *
+ * @buffer RNG data output
+ * @size RNG data size
+ */
+static TEE_Result imx_ele_rng_get_random(uint8_t *buffer, size_t size)
+{
+	uint64_t timeout = timeout_init_us(10 * 1000);
+	paddr_t pa = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct rng_get_random_cmd {
+		uint32_t addr_msb;
+		uint32_t addr_lsb;
+		uint32_t size;
+		uint32_t crc;
+	} cmd = { };
+	struct imx_mu_msg msg = {
+		.header.version = ELE_VERSION_HSM,
+		.header.size = SIZE_MSG_32(cmd),
+		.header.tag = ELE_REQUEST_TAG,
+		.header.command = ELE_CMD_RNG_GET,
+	};
+
+	assert(buffer && size);
```

Comment:
assert buffer and size are aligned on cache lines.

Reply:
no longer applies